src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfoInternalOfT.cs

```diff
@@ -93,5 +124,28 @@ private void SetCreateObjectFunc(Func<T>? createObjectFunc)
                 CreateObject = () => createObjectFunc();
             }
         }
+
+        private void DetermineIfCanUseSerializeFastPath()
+        {
+            Debug.Assert(Options == Options._context!.Options);
+            JsonSerializerOptions? contextDefaultOptions = Options._context!.DefaultOptions;
+            if (contextDefaultOptions == null)
+            {
+                throw new InvalidOperationException($"To specify a fast-path implementation for type {typeof(T)}, context {Options._context.GetType()} must specify default options.");
+            }
+
+            UseFastPathOnWrite =
+                // Guard against unsupported features
+                Options.Encoder == null &&
+                Options.NumberHandling == JsonNumberHandling.Strict &&
+                Options.ReferenceHandlingStrategy == ReferenceHandlingStrategy.None &&
+                // Ensure options values are consistent with expected defaults.
+                Options.DefaultIgnoreCondition == contextDefaultOptions.DefaultIgnoreCondition &&
+                Options.IgnoreReadOnlyFields == contextDefaultOptions.IgnoreReadOnlyFields &&
+                Options.IgnoreReadOnlyProperties == contextDefaultOptions.IgnoreReadOnlyProperties &&
+                Options.IncludeFields == contextDefaultOptions.IncludeFields &&
+                Options.PropertyNamingPolicy == contextDefaultOptions.PropertyNamingPolicy &&
+                Options.WriteIndented == contextDefaultOptions.WriteIndented;
+        }
```

Comment:
As discussed prior, for perf can we generate a single string to compare instead? The string would only need to be generated once on each side.

Reply:
I was able to make this check more efficient by moving it to the context class itself - https://github.com/dotnet/runtime/pull/53212/commits/1dd5d734dd66d163b48ed406bc71fc308c63e2eb. Now we only compute the value just once during the application: when the context class is initialized.

On generating strings to make the check quicker - I'll tackle this in a follow up.